Support for FreeBSD

## Updated proposal from 2017/9

Proposal (by @karelz - https://github.com/dotnet/corefx/issues/1626#issuecomment-329840518) will be updated in top-post based on further discussion and proposal changes.

We discussed community-driven port for FreeBSD with @RussellHaley (from FreeBSD community) and @wfurt (from .NET Core team) who both expressed interest in the work.
Here's a plan proposal we put together (feedback / suggestions are welcome):

1. Produce binaries in CoreCLR & CoreFX repo targeting FreeBSD - using hacks is fine
    * Hard to parallelize, @wfurt will work on that
    * The build can be mix of builds from other platforms (Mac, Linux) targeting FreeBSD
    * We will need documented steps (on FreeBSD wiki) to reproduce the build with FreeBSD-specific bug fixes
2. Run & stabilize CoreCLR tests (using corerun)
    * Tests may be built on another platform
    * Goal: Provides basic quality of runtime
3. Run & stabilize CoreFX tests (using corerun)
    * Tests may be built on another platform
    * Note this requires xunit. We believe, based on our past porting experience, once [2] is done, xunit will just work.
    * This can be in theory parallelized with [2] - it may require shortcutting xunit (e.g. generate static execution recipe on another platform)
    * We can expose new OSPlatform API for FreeBSD when the pass rate is reasonable: see dotnet/corefx#23989
4. Full stack build on FreeBSD (using corerun as bootstrapper from [1]-[3])
    * We will need all tools (nuget, msbuild, roslyn) to work on boostrapping .NET Core
5. Installers (FreeBSD ports)
    * First-stage: Using product binaries from nuget feeds
    * Second-stage: Build product from source (blocked on build from source effort)
    * Requires FreeBSD community expertise and guidance on design
    * Note: We can link FreeBSD packages also from official .NET Core download pages as community-support packages
6. Regular build and test runs on FreeBSD
    * Goal: Make sure changes in .NET Core repos breaking FreeBSD are known early
    * Design needed
    * Requires FreeBSD community expertise and guidance on design

Operation principles:
* Changes in [2]-[4] should be done primarily in CoreCLR/CoreFX repos (due to CLA signing requirements, code reviews from .NET Core team experts/members. etc.)
* We will track high-level work on this issue. Specific bugs will be filed as separate issues.

If anyone is interested in helping, please let us know here. We can easily distribute work items from [2] & [3] above once we are far enough with [1].

---------------------------------------------

## Original proposal from @ghuntley from 2015/5

This issue is to discuss unit(s) of work to actually produce FreeBSD assemblies for corefx.  

> @stephentoub - There's what's likely a more pressing issue, which is actually building for FreeBSD. Today, when we need to specialize an assembly for a particular platform, we effectively have three builds, producing three different managed assemblies: Windows, Linux, OSX. Sounds like at least for now we'll need a fourth, FreeBSD. I suggest you start by modifying the build to support an IsFreeBSD property (or just IsBSD of you think there's a high chance that the implementations across BSDs will be the same even with varied kernels) along with the appropriate OSGroup targets. That can then be used in the csproj files as needed to specialize an assembly with FreeBSD-specific code.

Related issue(s) 
- dotnet/runtime#14536 (OSGroup identifier in the public API)
- dotnet/runtime#14507 (OSGroup identifier in the private API)

/cc: @janhenke @josteink

I opted to use https://github.com/dotnet/runtime/pull/45664
Clr builds up to Clr.Tools subset then fails with
```
/root/runtime/.dotnet/sdk/5.0.100/Microsoft.Common.CurrentVersion.targets(4818,5): error MSB3030: Could not copy the file "/root/runtime/artifacts/bin/coreclr/FreeBSD.x64.Release/libjitinterface" because it was not found. [/root/runtime/src/coreclr/src/tools/aot/crossgen2/crossgen2.csproj]
/root/runtime/.dotnet/sdk/5.0.100/Microsoft.Common.CurrentVersion.targets(4818,5): error MSB3030: Could not copy the file "/root/runtime/artifacts/bin/coreclr/FreeBSD.x64.Release/libclrjit" because it was not found. [/root/runtime/src/coreclr/src/tools/aot/crossgen2/crossgen2.csproj]
```

subset "mono" and subset "libs" complete with no errors

@jasonpugsley that did the trick!
`/root/runtime/src/coreclr/src/tools/aot/crossgen2/crossgen2.csproj : error NU1101: Unable to find package Microsoft.AspNetCore.App.Runtime.freebsd-x64. No packages exist with this id in source(s):`
I knew I forgot to do something a few days ago! This should be interesting

of course it happens now. I need A to build A. Great use of product, terrible use of my sanity
`/root/aspnetcore/src/Tools/dotnet-watch/src/dotnet-watch.csproj : error NU1101: Unable to find package Microsoft.AspNetCore.App.Runtime.freebsd-x64. No packages exist with this id in source(s): darc-pub-dotnet-efcore-494e218, darc-pub-dotnet-runtime-b02e13a, dotnet-eng, dotnet-tools, dotnet31-transport, dotnet5, dotnet5-transport, localstore, nuget.org [/root/aspnetcore/.dotnet/sdk/5.0.100/NuGet.targets]`

for the runtime using all of https://github.com/dotnet/runtime/issues/14537#issuecomment-739709317 and a patch to add `-PDBFMT=PORTABLE` to the IlasmFlags everything builds up to Libs.Src where it fails with
```
Segmentation fault
/root/.nuget/packages/microsoft.net.sdk.il/5.0.0-preview.8.20359.4/targets/Microsoft.NET.Sdk.IL.targets(138,5): error MSB3073: The command "/root/.nuget/packages/runtime.freebsd-x64.microsoft.netcore.ilasm/5.0.1/runtimes/freebsd-x64/native/ilasm -QUIET -NOLOGO -OPTIMIZE -DLL  -DEBUG=OPT -PDBFMT=PORTABLE -INCLUDE=/root/runtime/artifacts/obj/System.Runtime.CompilerServices.Unsafe/netstandard2.0-Release/version/ -OUTPUT="/root/runtime/artifacts/obj/System.Runtime.CompilerServices.Unsafe/netstandard2.0-Release/System.Runtime.CompilerServices.Unsafe.dll" -KEY="/root/.nuget/packages/microsoft.dotnet.arcade.sdk/5.0.0-beta.20560.3/tools/snk/MSFT.snk" System.Runtime.CompilerServices.Unsafe.il" exited with code 139. [/root/runtime/src/libraries/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.ilproj]
```

Went back and built the full 3.1.11 SDK to feed into the requirements for ASPNetCore 5.0.x.
Everything works in the 3.1.11 leg! ASPNetCore 5.0.x still refuses to build

`/root/aspnetcore/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj : error NU1604: Project dependency Microsoft.NETCore.App.Runtime.freebsd-x64 does not contain an inclusive lower bound. Include a lower bound in the dependency version to ensure consistent restore results. [/root/aspnetcore/.dotnet/sdk/5.0.101/NuGet.targets]`

This error does not make sense given what I could find on NU1604. The only thing that I can think of is some file/version inconsistency occurred somewhere when building everything else for this.

I'm trying to build Radarr with this SDK but running into the following issue: 
```
/root/git_working/Radarr/src/NzbDrone.Console/Radarr.Console.csproj : error NU1101: Unable to find package Microsoft.AspNetCore.App.Runtime.freebsd-x64. No packages exist with this id in source(s): FluentMigrator, MyFeed, nuget.org, NUnit [/root/git_working/Radarr/src/Radarr.sln]
```
I guess this needs something from aspnetcore not included in your SDK bundle above?

Trying to native build runtime for dotNET 6.0-preview1 (using dotNET 5.0.202 SDK cross-compiled under Linux).
Command I've used `./build.sh -subset clr -os FreeBSD /p:CrossgenOutput=false /p:DISABLE_CROSSGEN=True`
Now the clr I think almost compiled, there are some weird errors and the end:

```  
  [100%] Built target runtimeinfo
Install the project...
  -- Install configuration: "DEBUG"
  Repo successfully built.
  Product binaries are available at /usr/home/sec/runtime/artifacts/bin/coreclr/FreeBSD.x64.Debug
  runincontext -> /usr/home/sec/runtime/artifacts/bin/coreclr/FreeBSD.x64.Debug/runincontext.dll
  ILCompiler.Reflection.ReadyToRun -> /usr/home/sec/runtime/artifacts/bin/coreclr/FreeBSD.x64.Debug/ILCompiler.Reflection.ReadyToRun.dll
  ILCompiler.Diagnostics -> /usr/home/sec/runtime/artifacts/bin/ILCompiler.Diagnostics/x64/Debug/ILCompiler.Diagnostics.dll
  R2RDump -> /usr/home/sec/runtime/artifacts/bin/coreclr/FreeBSD.x64.Debug/R2RDump/R2RDump.dll
  ILCompiler.TypeSystem.ReadyToRun -> /usr/home/sec/runtime/artifacts/bin/ILCompiler.TypeSystem.ReadyToRun/x64/Debug/ILCompiler.TypeSystem.ReadyToRun.dll
  dotnet-pgo -> /usr/home/sec/runtime/artifacts/bin/coreclr/FreeBSD.x64.Debug/dotnet-pgo/dotnet-pgo.dll
  R2RTest -> /usr/home/sec/runtime/artifacts/bin/coreclr/FreeBSD.x64.Debug/R2RTest/R2RTest.dll
  ILCompiler.DependencyAnalysisFramework -> /usr/home/sec/runtime/artifacts/bin/ILCompiler.DependencyAnalysisFramework/x64/Debug/ILCompiler.DependencyAnalysisFramework.dll
  ILCompiler.ReadyToRun -> /usr/home/sec/runtime/artifacts/bin/ILCompiler.ReadyToRun/x64/Debug/ILCompiler.ReadyToRun.dll
  crossgen2 -> /usr/home/sec/runtime/artifacts/bin/coreclr/FreeBSD.x64.Debug/crossgen2/crossgen2.dll
  CoreTestAssembly -> /usr/home/sec/runtime/artifacts/bin/CoreTestAssembly/x64/Debug/CoreTestAssembly.dll
ELF binary type "0" not known.
  /tmp/tmp06262290b1b448348b21bb0b6cd8965a.exec.cmd: /home/sec/.nuget/packages/runtime.linux-x64.microsoft.netcore.ilasm/6.0.0-alpha.1.21068.2/runtimes/linux-x64/native/ilasm: Exec format error
/home/sec/.nuget/packages/microsoft.net.sdk.il/6.0.0-alpha.1.21068.2/targets/Microsoft.NET.Sdk.IL.targets(145,5): error MSB3073: The command ""/home/sec/.nuget/packages/runtime.linux-x64.microsoft.netcore.ilasm/6.0.0-alpha.1.21068.2/runtimes/linux-x64/native/ilasm" -QUIET -NOLOGO -DLL  -OUTPUT="/usr/home/sec/runtime/artifacts/obj/coreclr/ILTestAssembly/x64/Debug/ILTestAssembly.dll"  Main.il ILDisassembler.il InstanceFieldLayout.il StaticFieldLayout.il VirtualFunctionOverride.il Signature.il MethodImplOverride1.il" exited with code 126. [/usr/home/sec/runtime/src/coreclr/tools/aot/ILCompiler.TypeSystem.ReadyToRun.Tests/ILTestAssembly/ILTestAssembly.ilproj]

Build FAILED.
```
Why it's using linux runtime for something?

Then I tried to build `libs`, not it's not that simple - a lot of `error NU1012: Platform version is not present for one or more target frameworks, even though they have specified a platform: net6.0-linux, net6.0-osx, net6.0-freebsd, net6.0-ios, net6.0-tvos, net6.0-browser, net6.0-illumos, net6.0-solaris` errors - how to approach/fix that?

After that I tried the same thing, but with `v5.0.5` tag, this time using clang10 and some changes from 
@jasonpugsley https://github.com/jasonpugsley/runtime/commit/cc43bef7f1322871f2e538d41b8263bf958f40a4

For clr:
```

  Product binaries are available at /usr/home/sec/runtim2/artifacts/bin/coreclr/FreeBSD.x64.Debug
  ILCompiler.DependencyAnalysisFramework -> /usr/home/sec/runtim2/artifacts/bin/ILCompiler.DependencyAnalysisFramework/x64/Debug/ILCompiler.DependencyAnalysisFramework.dll
  ILCompiler.TypeSystem.ReadyToRun -> /usr/home/sec/runtim2/artifacts/bin/ILCompiler.TypeSystem.ReadyToRun/x64/Debug/ILCompiler.TypeSystem.ReadyToRun.dll
  ILCompiler.ReadyToRun -> /usr/home/sec/runtim2/artifacts/bin/ILCompiler.ReadyToRun/x64/Debug/ILCompiler.ReadyToRun.dll
/usr/home/sec/dotnet_from_sec_build/sdk/5.0.202/Microsoft.Common.CurrentVersion.targets(4919,5): error MSB3030: Could not copy the file "/usr/home/sec/runtim2/artifacts/bin/coreclr/FreeBSD.x64.Debug/libjitinterface" because it was not found. [/usr/home/sec/runtim2/src/coreclr/src/tools/aot/crossgen2/crossgen2.csproj]
/usr/home/sec/dotnet_from_sec_build/sdk/5.0.202/Microsoft.Common.CurrentVersion.targets(4919,5): error MSB3030: Could not copy the file "/usr/home/sec/runtim2/artifacts/bin/coreclr/FreeBSD.x64.Debug/libclrjit" because it was not found. [/usr/home/sec/runtim2/src/coreclr/src/tools/aot/crossgen2/crossgen2.csproj]

Build FAILED.
```
I didn't took all the changes from @jasonpugsley , but I assume that it's fixable.
but `libs` managed to build without any errors.

@joperator re-adding the removed cert fixed the issue thanks!

- platform detection(?)

This issue is actually to net6 and didn't exist in net5 or with 3.1...
I am still trying to figure out why runtime still thinks its suddenly on Linux during the final part of the coreclr subset.

manually building: `src/coreclr/.nuget/` correctly creates freebsd named packages (`runtime.freebsd-x64.Microsoft.NETCore.ILAsm.6.0.0-dev.21281.99.nupkg`) 
building with the build script still results in it looking for and trying to produce linux named items.

the problem seems to be for only a few items:
```
/root/runtime/src/coreclr/tools/aot/ILCompiler.TypeSystem.ReadyToRun.Tests/ILTestAssembly/ILTestAssembly.ilproj : error NU1603: ILTestAssembly depends on runtime.linux-x64.Microsoft.NETCore.ILAsm (>= 6.0.0-dev.21281.99) but runtime.linux-x64.Microsoft.NETCore.ILAsm 6.0.0-dev.21281.99 was not found. An approximate best match of runtime.linux-x64.Microsoft.NETCore.ILAsm 6.0.0-preview.1.21071.15 was resolved. [/root/runtime/Build.proj]
/root/runtime/src/libraries/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.ilproj : error NU1603: System.Runtime.CompilerServices.Unsafe depends on runtime.linux-x64.Microsoft.NETCore.ILAsm (>= 6.0.0-dev.21281.99) but runtime.linux-x64.Microsoft.NETCore.ILAsm 6.0.0-dev.21281.99 was not found. An approximate best match of runtime.linux-x64.Microsoft.NETCore.ILAsm 6.0.0-preview.1.21071.15 was resolved. [/root/runtime/Build.proj]
```
When I get some time to deep-dive it, I will look at the binlog and see if that has any clues

- ibcmerge

if `/p:EnableNgenOptimization=false` is not included in the build then the libs subset fails with a bunch of
`The application to execute does not exist: '/root/runtime/.packages/microsoft.dotnet.ibcmerge/5.0.7-beta.20159.1/tools/netcoreapp2.0/ibcmerge.dll'`) 

I can't find anything that restores that package so im outta ideas on that one :(
maybe its related to the first issue? I really have no idea

Another update and it doesn't look good :(

First item:
The runtime build process fails if doing a full run but has no issues just doing `coreclr` subset first then the rest of the subsets. The crash that I am encountering during a full run is this ugly mess:
```

/root/runtime/dotnet.sh /root/runtime/artifacts/bin/coreclr/FreeBSD.x64.Release/crossgen2/crossgen2.dll -o:/root/runtime/artifacts/bin/coreclr/FreeBSD.x64.Release/System.Private.CoreLib.dll -r:/root/runtime/artifacts/bin/coreclr/FreeBSD.x64.Release/IL/*.dll --targetarch:x64 -O /root/runtime/artifacts/bin/coreclr/FreeBSD.x64.Release/IL/System.Private.CoreLib.dll
EXEC : error : One or more errors occurred. (External component has thrown an exception.) (External component has thrown an exception.) [/root/runtime/src/coreclr/crossgen-corelib.proj]
  System.AggregateException: One or more errors occurred. (External component has thrown an exception.) (External component has thrown an exception.)
   ---> System.Runtime.InteropServices.SEHException (0x80004005): External component has thrown an exception.
     at System.Runtime.InteropServices.NativeLibrary.LoadByName(String libraryName, QCallAssembly callingAssembly, Boolean hasDllImportSearchPathFlag, UInt32 dllImportSearchPathFlag, Boolean throwOnError)
     at System.Runtime.InteropServices.NativeLibrary.LoadLibraryByName(String libraryName, Assembly assembly, Nullable`1 searchPath, Boolean throwOnError)
     at System.Runtime.InteropServices.NativeLibrary.Load(String libraryName, Assembly assembly, Nullable`1 searchPath)
     at Internal.JitInterface.JitConfigProvider.<>c__DisplayClass5_0.<Initialize>b__0(String libName, Assembly assembly, Nullable`1 searchPath)
     at System.Runtime.InteropServices.NativeLibrary.LoadLibraryCallbackStub(String libraryName, Assembly assembly, Boolean hasDllImportSearchPathFlags, UInt32 dllImportSearchPathFlags)
     at Internal.JitInterface.CorInfoImpl.getJit()
     at Internal.JitInterface.CorInfoImpl..ctor()
     at Internal.JitInterface.CorInfoImpl..ctor(ReadyToRunCodegenCompilation compilation)
     at ILCompiler.ReadyToRunCodegenCompilation.<ComputeDependencyNodeDependencies>b__35_1(Thread thread)
     at System.Runtime.CompilerServices.ConditionalWeakTable`2.GetValueLocked(TKey key, CreateValueCallback createValueCallback)
     at System.Runtime.CompilerServices.ConditionalWeakTable`2.GetValue(TKey key, CreateValueCallback createValueCallback)
     at ILCompiler.ReadyToRunCodegenCompilation.<ComputeDependencyNodeDependencies>b__35_0(DependencyNodeCore`1 dependency)
     at System.Threading.Tasks.Parallel.<>c__DisplayClass33_0`2.<ForEachWorker>b__0(Int32 i)
     at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
  --- End of stack trace from previous location ---
     at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
     at System.Threading.Tasks.TaskReplicator.Replica`1.ExecuteAction(Boolean& yieldedBeforeCompletion)
     at System.Threading.Tasks.TaskReplicator.Replica.Execute()
     --- End of inner exception stack trace ---
     at System.Threading.Tasks.TaskReplicator.Run[TState](ReplicatableUserAction`1 action, ParallelOptions options, Boolean stopOnFirstFailure)
     at System.Threading.Tasks.Parallel.ForWorker[TLocal](Int32 fromInclusive, Int32 toExclusive, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Func`4 bodyWithLocal, Func`1 localInit, Action`1 localFinally)
  --- End of stack trace from previous location ---
     at System.Threading.Tasks.Parallel.ThrowSingleCancellationExceptionOrOtherException(ICollection exceptions, CancellationToken cancelToken, Exception otherException)
     at System.Threading.Tasks.Parallel.ForWorker[TLocal](Int32 fromInclusive, Int32 toExclusive, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Func`4 bodyWithLocal, Func`1 localInit, Action`1 localFinally)
     at System.Threading.Tasks.Parallel.ForEachWorker[TSource,TLocal](IList`1 list, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWithEverything, Func`1 localInit, Action`1 localFinally)
     at System.Threading.Tasks.Parallel.ForEachWorker[TSource,TLocal](IEnumerable`1 source, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWithEverything, Func`1 localInit, Action`1 localFinally)
     at System.Threading.Tasks.Parallel.ForEach[TSource](IEnumerable`1 source, ParallelOptions parallelOptions, Action`1 body)
     at ILCompiler.ReadyToRunCodegenCompilation.ComputeDependencyNodeDependencies(List`1 obj)
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ComputeMarkedNodes()
     at ILCompiler.ReadyToRunCodegenCompilation.Compile(String outputFile)
     at ILCompiler.Program.Run(String[] args)
     at ILCompiler.Program.Main(String[] args)
   ---> (Inner Exception #1) System.Runtime.InteropServices.SEHException (0x80004005): External component has thrown an exception.
     at System.Runtime.InteropServices.NativeLibrary.LoadByName(String libraryName, QCallAssembly callingAssembly, Boolean hasDllImportSearchPathFlag, UInt32 dllImportSearchPathFlag, Boolean throwOnError)
     at System.Runtime.InteropServices.NativeLibrary.LoadLibraryByName(String libraryName, Assembly assembly, Nullable`1 searchPath, Boolean throwOnError)
     at System.Runtime.InteropServices.NativeLibrary.Load(String libraryName, Assembly assembly, Nullable`1 searchPath)
     at Internal.JitInterface.JitConfigProvider.<>c__DisplayClass5_0.<Initialize>b__0(String libName, Assembly assembly, Nullable`1 searchPath)
     at System.Runtime.InteropServices.NativeLibrary.LoadLibraryCallbackStub(String libraryName, Assembly assembly, Boolean hasDllImportSearchPathFlags, UInt32 dllImportSearchPathFlags)
     at Internal.JitInterface.CorInfoImpl.getJit()
     at Internal.JitInterface.CorInfoImpl..ctor()
     at Internal.JitInterface.CorInfoImpl..ctor(ReadyToRunCodegenCompilation compilation)
     at ILCompiler.ReadyToRunCodegenCompilation.<ComputeDependencyNodeDependencies>b__35_1(Thread thread)
     at System.Runtime.CompilerServices.ConditionalWeakTable`2.GetValueLocked(TKey key, CreateValueCallback createValueCallback)
     at System.Runtime.CompilerServices.ConditionalWeakTable`2.GetValue(TKey key, CreateValueCallback createValueCallback)
     at ILCompiler.ReadyToRunCodegenCompilation.<ComputeDependencyNodeDependencies>b__35_0(DependencyNodeCore`1 dependency)
     at System.Threading.Tasks.Parallel.<>c__DisplayClass33_0`2.<ForEachWorker>b__0(Int32 i)
     at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
  --- End of stack trace from previous location ---
     at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
     at System.Threading.Tasks.TaskReplicator.Replica`1.ExecuteAction(Boolean& yieldedBeforeCompletion)
     at System.Threading.Tasks.TaskReplicator.Replica.Execute()<---
  
/root/runtime/src/coreclr/crossgen-corelib.proj(105,5): error MSB3073: The command "/root/runtime/dotnet.sh /root/runtime/artifacts/bin/coreclr/FreeBSD.x64.Release/crossgen2/crossgen2.dll -o:/root/runtime/artifacts/bin/coreclr/FreeBSD.x64.Release/System.Private.CoreLib.dll -r:/root/runtime/artifacts/bin/coreclr/FreeBSD.x64.Release/IL/*.dll --targetarch:x64 -O /root/runtime/artifacts/bin/coreclr/FreeBSD.x64.Release/IL/System.Private.CoreLib.dll" exited with code 1.

```

Second item: 
Still getting segfaults from the final product. I have attached the output of `truss -f -H -o trussdotnet.log ./dotnet --info` for someone who has a better understanding of what might be going on.

[trussdotnet.log](https://github.com/dotnet/runtime/files/6602064/trussdotnet.log)


If I have some free time this weekend I will try an earlier preview/commit and hopefully be able to get a better idea of when this started to fail.

@Thefrank would you mind sharing steps/scripts that you're using to build v6 previews? After many tries I cannot get them to build in sync. Runtime is fine, but aspnetcore is failing, for ex.:
```
/home/sec/dotnet-freebsd-crossbuild/aspnetcore/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj : error NU1603: Microsoft.AspNetCore.App.Runtime.freebsd-x64 depends on Microsoft.NETCore.App.Runtime.freebsd-x64 (>= 6.0.0-preview.2.21154.6) but Microsoft.NETCore.App.Runtime.freebsd-x64 6.0.0-preview.2.21154.6 was not found. An approximate best match of Microsoft.NETCore.App.Runtime.freebsd-x64 6.0.0-preview.2.21308.99 was resolved. [/home/sec/dotnet-freebsd-crossbuild/aspnetcore/.dotnet/sdk/6.0.100-preview.2.21118.6/NuGet.targets]
```

I think I've adjusted your build script for v6 and tried those tags:
```
RUNTIMETAG=v6.0.0-preview.2.21154.6
ASPNETTAG=v6.0.0-preview.2.21154.6
INSTALLERTAG=v6.0.100-preview.2.21155.3
```

If I could get reproducable, automated build I could help with finding commit that broke the runtime - at least I hope so :)

@wfurt it does not build for me:

- `gen-buildsys-clang.sh` does not appear to be able to understand that llvm versions > 9 exist. This is not a huge problem, llvm9->12 are supported under FreeBSD 12
- `LLDB_INCLUDE_DIR` `LLDB_LIB_DIR` need to be set otherwise the configure fails

manually making the above changes gets the build to at least start but it still fails with:
```
/root/diagnostics/src/pal/src/misc/sysinfo.cpp:301:10: error: use of undeclared identifier 'sysctlnametomib'
    rc = sysctlnametomib("vm.swap_info", mib, &length);
         ^
/root/diagnostics/src/pal/src/misc/sysinfo.cpp:309:18: error: use of undeclared identifier 'sysctl'
            rc = sysctl(mib, 3, &xsw, &length, NULL, 0);
```

There are other annoying (but non-critical) things:
- despite restore succeeding it still reports that it failed with error code 139. using `--ci` seems to override this behavior and the build continues
- it seems to default to llvm 3.9. This is not a problem if 100% building from source or using an older docker/vm but many OSes don't have packages for this due to age.
- `eng/build.sh` doesn't understand that amd64 = x64 as `uname -m` and `uname -p` both return `amd64` on FreeBSD. It thankfully defaults to x64 without needing to change case logic.
- cpu number dectection is wrong for freebsd thus defaulting to `-j 1`. Not much of an issue on a small project


I am likely missing something obvious to get this fixed but I am feeling very overwhelmed by everything that needs to be done in order to get anything dotnet6 running under FreeBSD.

@wfurt sadly, it will likely need quite a bit of work if `elfreader` is to be included

```
[ 78%] Building CXX object src/SOS/dbgutil/CMakeFiles/dbgutil.dir/elfreader.cpp.o
/root/diagnostics/src/SOS/dbgutil/elfreader.cpp:396:33: error: cannot initialize a parameter of type 'void *' with an rvalue of type 'const char *'
                if (!ReadMemory(map.l_name + i, &ch, sizeof(ch))) {
                                ^~~~~~~~~~~~~~
/root/diagnostics/src/SOS/dbgutil/elfreader.h:68:35: note: passing argument to parameter 'address' here
    virtual bool ReadMemory(void* address, void* buffer, size_t size) = 0;
                                  ^
/root/diagnostics/src/SOS/dbgutil/elfreader.cpp:409:21: error: cannot initialize a parameter of type 'uint64_t' (aka 'unsigned long') with an lvalue of type 'caddr_t' (aka 'char *')
        VisitModule(map.l_addr, moduleName);
                    ^~~~~~~~~~
/root/diagnostics/src/SOS/dbgutil/elfreader.h:66:39: note: passing argument to parameter 'baseAddress' here
    virtual void VisitModule(uint64_t baseAddress, std::string& moduleName) { };
                                      ^
2 errors generated.
*** [src/SOS/dbgutil/CMakeFiles/dbgutil.dir/elfreader.cpp.o] Error code 1
```

The new wall I am hitting: other OSes define `Link_map` in their `link.h` differently than the Linux kernel does. 

Spoiler: of the six platforms listed, 5 of the 6 are all different from eachother.


Linux: https://man7.org/linux/man-pages/man3/dlinfo.3.html 
FreeBSD: https://www.freebsd.org/cgi/man.cgi?query=dlinfo&sektion=3 

and just for insanity if trying to make it multiplatform:
NetBSD: https://man.netbsd.org/dlinfo.3
OpenBSD (dlinfo is not in OpenBSD AFAIK): https://man.openbsd.org/OpenBSD-5.3/link.5
(These two are actually the same!)
Illumos: https://illumos.org/man/3C/dlinfo
SunOS: https://docs.oracle.com/cd/E36784_01/html/E36874/dlinfo-3c.html

`src/SOS/dbgutil/elfreader.h` and `elfreader.cpp` are very much Linux-only. The rest of `dbgutil` looks rather agnostic, is needed to build the repo, and can build around `elfreader`...so I built around `elfreader` 

Final output is three files:
```
libsos.so       
libsosplugin.so 
sosdocsunix.txt
```
Victory?! No.

```
(lldb) plugin load diagnostics/artifacts/bin/dotnet-sos/Debug/netcoreapp2.1/publish/freebsd.12-x64/libsosplugin.so
error: this file does not represent a loadable dylib
```

/HEADDESK/

Nice to see some progress :) I've checked build provided by @ta264 and run some console and mvc app with success.

Going back to native build (preview5 using preview5 sdk) on FreeBSD, I was able to build clr subset  with some changes (it still try to use ilasm from linux, but doing replace of the linux binary with FreeBSD one did the trick, along with some changes to crossgen2 to support TargetOS of FreeBSD.

Trying to build libs subset, resulted in lots of errors like:
`
/usr/home/sec/dotnet6p5/sdk/6.0.100-preview.5.21302.13/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.ImportWorkloads.targets(33,5): error NETSDK1147: To build this project, the following workloads must be installed: microsoft-tvos-sdk-full [/usr/home/sec/runtime/Build.proj]
/usr/home/sec/dotnet6p5/sdk/6.0.100-preview.5.21302.13/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.ImportWorkloads.targets(33,5): error NETSDK1147: To install these workloads, run the following command: dotnet workload install microsoft-tvos-sdk-full [/usr/home/sec/runtime/Build.proj]
`

mono subset is not supported at all (even after some changes to src/mono/CMakeLists.txt to support both FreeBSD and amd64) it fails:
`
Running ' cmake --build . --target install --config Release --parallel 1' in '/usr/home/sec/runtime/artifacts/obj/mono/FreeBSD.x64.Release/'
  [  1%] Building C object mono/mini/CMakeFiles/monosgen-objects.dir/__/metadata/w32event-unix.c.o
  In file included from /usr/home/sec/runtime/src/mono/mono/metadata/w32event-unix.c:11:
  In file included from /usr/home/sec/runtime/src/mono/mono/metadata/w32event.h:9:
  In file included from /usr/home/sec/runtime/src/mono/mono/mini/../eglib/glib.h:29:
  /usr/home/sec/runtime/artifacts/obj/mono/FreeBSD.x64.Release/mono/mini/../../mono/eglib/eglib-config.h:36:6: error: invalid token at start of a preprocessor expression
  #if  == 1
       ^
  1 error generated.
  *** [mono/mini/CMakeFiles/monosgen-objects.dir/__/metadata/w32event-unix.c.o] Error code 1
`

Tried using clang 9 and 10 - same effect.

> @Thefrank Maybe subtle ABI incompatibilities may be the cause of the crashes you have been getting with multi-core if the build process retrieves 11.x binaries to run them on 12.x.

Did native build under FreeBSD 13 using crossbuild as starter, then tried to build runtime with native build - still faulting, sometimes it's core dumped, sometimes it's diffrent error, like:
```
MSBUILD : error MSB1025: An internal failure occurred while running MSBuild.
Microsoft.Build.Shared.InternalErrorException: MSB0001: Internal MSBuild Error: Node 3 does not have a provider.
   at Microsoft.Build.CommandLine.MSBuildApp.BuildProject(String projectFile, String[] targets, String toolsVersion, Dictionary`2 globalProperties, Dictionary`2 restoreProperties, ILogger[] loggers, LoggerVerbosity verbosity, DistributedLoggerRecord[] distributedLoggerRecords, Int32 cpuCount, Boolean enableNodeReuse, TextWriter preprocessWriter, TextWriter targetsWriter, Boolean detailedSummary, ISet`1 warningsAsErrors, ISet`1 warningsAsMessages, Boolean enableRestore, ProfilerLogger profilerLogger, Boolean enableProfiler, Boolean interactive, Boolean isolateProjects, GraphBuildOptions graphBuildOptions, Boolean lowPriority, String[] inputResultsCaches, String outputResultsCache) in MSBuild.dll:token 0x600008d+0x5b2
   at Microsoft.Build.CommandLine.MSBuildApp.Execute(String[] commandLine) in MSBuild.dll:token 0x600008a+0x1f5
Unhandled exception: Microsoft.Build.Shared.InternalErrorException: MSB0001: Internal MSBuild Error: Node 3 does not have a provider.
   at Microsoft.Build.CommandLine.MSBuildApp.BuildProject(String projectFile, String[] targets, String toolsVersion, Dictionary`2 globalProperties, Dictionary`2 restoreProperties, ILogger[] loggers, LoggerVerbosity verbosity, DistributedLoggerRecord[] distributedLoggerRecords, Int32 cpuCount, Boolean enableNodeReuse, TextWriter preprocessWriter, TextWriter targetsWriter, Boolean detailedSummary, ISet`1 warningsAsErrors, ISet`1 warningsAsMessages, Boolean enableRestore, ProfilerLogger profilerLogger, Boolean enableProfiler, Boolean interactive, Boolean isolateProjects, GraphBuildOptions graphBuildOptions, Boolean lowPriority, String[] inputResultsCaches, String outputResultsCache) in MSBuild.dll:token 0x600008d+0x5b2
   at Microsoft.Build.CommandLine.MSBuildApp.Execute(String[] commandLine) in MSBuild.dll:token 0x600008a+0x1f5
   at Microsoft.Build.CommandLine.MSBuildApp.Main(String[] args) in MSBuild.dll:token 0x6000086+0x29
   at Microsoft.DotNet.Cli.Utils.MSBuildForwardingAppWithoutLogging.ExecuteInProc(String[] arguments) in Microsoft.DotNet.Cli.Utils.dll:token 0x6000106+0x62
Build failed with exit code 82. Check errors above.
##vso[task.complete result=Failed;]msbuild execution failed.
```

> @sec so it looks like the issue is present in all current versions of FreeBSD (11,12,13). Does `-maxcpucount:1` fix the issue?

Yes, that's the fix to get the native build done all the way up, as you wrote before. Did you do any changes to get the test running? I tried but it fails with:
```
Build FAILED.

/usr/home/sec/runtime/.dotnet/sdk/6.0.100-preview.6.21355.2/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(127,5): error NETSDK1084: There is no application host available for the specified RuntimeIdentifier 'freebsd-x64'. [/usr/home/sec/runtime/src/installer/tests/Assets/TestProjects/StandaloneApp3x/StandaloneApp3x.csproj]
/usr/home/sec/runtime/.dotnet/sdk/6.0.100-preview.6.21355.2/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(127,5): error NETSDK1084: There is no application host available for the specified RuntimeIdentifier 'freebsd-x64'. [/usr/home/sec/runtime/src/installer/tests/Assets/TestProjects/StandaloneApp3x/StandaloneApp3x.csproj]
/usr/home/sec/runtime/.dotnet/sdk/6.0.100-preview.6.21355.2/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(127,5): error NETSDK1084: There is no application host available for the specified RuntimeIdentifier 'freebsd-x64'. [/usr/home/sec/runtime/src/installer/tests/Assets/TestProjects/StandaloneApp3x/StandaloneApp3x.csproj]
/usr/home/sec/runtime/.dotnet/sdk/6.0.100-preview.6.21355.2/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(127,5): error NETSDK1084: There is no application host available for the specified RuntimeIdentifier 'freebsd-x64'. [/usr/home/sec/runtime/src/installer/tests/Assets/TestProjects/StandaloneApp3x/StandaloneApp3x.csproj]
/usr/home/sec/runtime/src/installer/tests/PrepareTestAssets/PrepareTestAssets.proj(123,5): error MSB4181: The "MSBuild" task returned false but did not log an error.
    0 Warning(s)
    5 Error(s)
```
even though SDK is able to create new projects, build and run them?

What is your setup for the linux VM? (distro, packages). I am getting failures on the coreclr part using a crossrootfs (FreeBSD 12.2 or FreeBSD 13.0) built via the arcade script.

```
  /home/frank/runtime/src/coreclr/debug/daccess/daccess.cpp:7243:38: error: invalid suffix 'x' on integer constant
      *dacTableAddress = baseAddress + DAC_TABLE_RVA;
                                       ^
  /home/frank/runtime/artifacts/obj/coreclr/FreeBSD.x64.Debug/inc/dactablerva.h:1:24: note: expanded from macro 'DAC_TABLE_RVA'
  #define DAC_TABLE_RVA 0x          e25a00
                         ^
  1 error generated.
  make[2]: *** [debug/daccess/CMakeFiles/daccess.dir/build.make:124: debug/daccess/CMakeFiles/daccess.dir/daccess.cpp.o] Error 1
  make[2]: *** Waiting for unfinished jobs....
  make[1]: *** [CMakeFiles/Makefile2:4498: debug/daccess/CMakeFiles/daccess.dir/all] Error 2
  make: *** [Makefile:130: all] Error 2
```

The docker image (FreeBSD 12.1) gets a bit further, into mono but still errors out:
```
In file included from /runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_calendarData.c:8:
  In file included from /runtime/src/mono/mono/mini/../../../libraries/Native/Unix/System.Globalization.Native/pal_locale_internal.h:6:
  /runtime/src/mono/mono/mini/../../../libraries/Native/Unix/System.Globalization.Native/pal_icushim_internal.h:22:10: fatal error: 'unicode/ucurr.h' file not found
  #include <unicode/ucurr.h>
           ^~~~~~~~~~~~~~~~~
  1 error generated.
  mono/mini/CMakeFiles/monosgen-objects.dir/build.make:2866: recipe for target 'mono/mini/CMakeFiles/monosgen-objects.dir/runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_calendarData.c.o' failed
  make[2]: *** [mono/mini/CMakeFiles/monosgen-objects.dir/runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_calendarData.c.o] Error 1
  make[2]: *** Waiting for unfinished jobs....
  [ 80%] Building C object mono/mini/CMakeFiles/monosgen-objects.dir/runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_idna.c.o
  In file included from /runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_casing.c:9:
  /runtime/src/mono/mono/mini/../../../libraries/Native/Unix/System.Globalization.Native/pal_icushim_internal.h:22:10: fatal error: 'unicode/ucurr.h' file not found
  #include <unicode/ucurr.h>
           ^~~~~~~~~~~~~~~~~
  1 error generated.
  In file included from /runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c:11:
  In file included from /runtime/src/mono/mono/mini/../../../libraries/Native/Unix/System.Globalization.Native/pal_errors_internal.h:6:
  /runtime/src/mono/mono/mini/../../../libraries/Native/Unix/System.Globalization.Native/pal_icushim_internal.h:22:10: fatal error: 'unicode/ucurr.h' file not found
  #include <unicode/ucurr.h>
           ^~~~~~~~~~~~~~~~~
  make[2]: *** [mono/mini/CMakeFiles/monosgen-objects.dir/runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_casing.c.o] Error 1
  mono/mini/CMakeFiles/monosgen-objects.dir/build.make:2880: recipe for target 'mono/mini/CMakeFiles/monosgen-objects.dir/runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_casing.c.o' failed
  1 error generated.
  mono/mini/CMakeFiles/monosgen-objects.dir/build.make:2894: recipe for target 'mono/mini/CMakeFiles/monosgen-objects.dir/runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c.o' failed
  make[2]: *** [mono/mini/CMakeFiles/monosgen-objects.dir/runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c.o] Error 1
  In file included from /runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_idna.c:7:
  /runtime/src/mono/mono/mini/../../../libraries/Native/Unix/System.Globalization.Native/pal_icushim_internal.h:22:10: fatal error: 'unicode/ucurr.h' file not found
  #include <unicode/ucurr.h>
           ^~~~~~~~~~~~~~~~~
  1 error generated.
  mono/mini/CMakeFiles/monosgen-objects.dir/build.make:2908: recipe for target 'mono/mini/CMakeFiles/monosgen-objects.dir/runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_idna.c.o' failed
  make[2]: *** [mono/mini/CMakeFiles/monosgen-objects.dir/runtime/src/libraries/Native/Unix/System.Globalization.Native/pal_idna.c.o] Error 1
  CMakeFiles/Makefile2:220: recipe for target 'mono/mini/CMakeFiles/monosgen-objects.dir/all' failed
  make[1]: *** [mono/mini/CMakeFiles/monosgen-objects.dir/all] Error 2
  Makefile:135: recipe for target 'all' failed
  make: *** [all] Error 2
/runtime/src/mono/mono.proj(500,5): error MSB3073: The command "TARGET_BUILD_ARCH=x64 PKG_CONFIG_PATH=/crossrootfs/x64/host/libdata/pkgconfig:/crossrootfs/x64/usr/libdata/pkgconfig:/crossrootfs/x64/usr/local/libdata/pkgconfig cmake --build . --target install --config Debug --parallel 16" exited with code 2.
```

> ```
>   /home/frank/runtime/src/coreclr/debug/daccess/daccess.cpp:7243:38: error: invalid suffix 'x' on integer constant
>       *dacTableAddress = baseAddress + DAC_TABLE_RVA;
>                                        ^
>   /home/frank/runtime/artifacts/obj/coreclr/FreeBSD.x64.Debug/inc/dactablerva.h:1:24: note: expanded from macro 'DAC_TABLE_RVA'
>   #define DAC_TABLE_RVA 0x          e25a00
>                          ^
> ```
According to `src/coreclr/pal/tools/gen-dactable-rva.sh` that hex number is meant to be zero padded but in your case it's being padded with spaces. It should be `0x0000000000e25a00`

This is what I get in my generated file:
`#define DAC_TABLE_RVA 0x0000000000dde600`

--- a/src/libraries/Common/src/Interop/FreeBSD/Interop.Process.GetProcInfo.cs
+++ b/src/libraries/Common/src/Interop/FreeBSD/Interop.Process.GetProcInfo.cs
@@ -186,11 +186,6 @@ internal static partial class Interop
         public static unsafe kinfo_proc* GetProcInfo(int pid, bool threads, out int count)
         {
             Span<int> sysctlName = stackalloc int[4];
-            int bytesLength = 0;
-            byte* pBuffer = null;
-            kinfo_proc* kinfo = null;
-
-            count = -1;
 
             if (pid == 0)
             {
@@ -207,23 +202,17 @@ internal static partial class Interop
             sysctlName[1] = KERN_PROC;
             sysctlName[0] = CTL_KERN;
 
-            try
-            {
-                Interop.Sys.Sysctl(sysctlName, ref pBuffer, ref bytesLength);
-                kinfo = (kinfo_proc*)pBuffer;
-                if (kinfo->ki_structsize != sizeof(kinfo_proc))
-                {
-                    // failed consistency check
-                    throw new ArgumentOutOfRangeException(nameof(pid));
-                }
+            byte* pBuffer = null;
+            int bytesLength = 0;
+            Interop.Sys.Sysctl(sysctlName, ref pBuffer, ref bytesLength);
 
-                count = (int)bytesLength / sizeof(kinfo_proc);
-            }
-            finally
-            {
-                Marshal.FreeHGlobal((IntPtr)pBuffer);
-            }
+            kinfo_proc* kinfo = (kinfo_proc*)pBuffer;
 
+            Debug.Assert(kinfo->ki_structsize == sizeof(kinfo_proc));
+
+            count = (int)bytesLength / sizeof(kinfo_proc);
+
+            // Buffer ownership transferred to the caller
             return kinfo;
         }
     }

--- a/src/libraries/Common/src/Interop/FreeBSD/Interop.Process.GetProcInfo.cs
+++ b/src/libraries/Common/src/Interop/FreeBSD/Interop.Process.GetProcInfo.cs
@@ -2,6 +2,7 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
+using System.Diagnostics;
 using System.Runtime.InteropServices;
 
 #pragma warning disable CA1823 // analyzer incorrectly flags fixed buffer length const (https://github.com/dotnet/roslyn/issues/37593)

--- a/src/libraries/Common/src/Interop/FreeBSD/Interop.Process.cs
+++ b/src/libraries/Common/src/Interop/FreeBSD/Interop.Process.cs
@@ -65,7 +65,7 @@ internal static partial class Interop
             }
             finally
             {
-                Marshal.FreeHGlobal((IntPtr)entries);
+                NativeMemory.Free(entries);
             }
         }
 
@@ -87,7 +87,7 @@ internal static partial class Interop
             }
             finally
             {
-                Marshal.FreeHGlobal((IntPtr)pBuffer);
+                NativeMemory.Free(pBuffer);
             }
         }
 
@@ -107,9 +107,9 @@ internal static partial class Interop
                 throw new ArgumentOutOfRangeException(nameof(pid));
             }
 
-            kinfo_proc* kinfo = GetProcInfo(pid, true, out int count);
             ProcessInfo info;
 
+            kinfo_proc* kinfo = GetProcInfo(pid, true, out int count);
             try
             {
                 if (count < 1)
@@ -142,7 +142,7 @@ internal static partial class Interop
             }
             finally
             {
-                Marshal.FreeHGlobal((IntPtr)kinfo);
+                NativeMemory.Free(kinfo);
             }
 
             return info;
@@ -160,12 +160,11 @@ internal static partial class Interop
         public static unsafe proc_stats GetThreadInfo(int pid, int tid)
         {
             proc_stats ret = default;
-            kinfo_proc* info = null;
             int count;
 
+            kinfo_proc* info = GetProcInfo(pid, (tid != 0), out count);
             try
             {
-                info = GetProcInfo(pid, (tid != 0), out count);
                 if (info != null && count >= 1)
                 {
                     if (tid == 0)
@@ -194,7 +193,7 @@ internal static partial class Interop
             }
             finally
             {
-                Marshal.FreeHGlobal((IntPtr)info);
+                NativeMemory.Free(info);
             }
 
             return ret;

--- a/src/libraries/System.Private.CoreLib/src/System/Environment.FreeBSD.cs
+++ b/src/libraries/System.Private.CoreLib/src/System/Environment.FreeBSD.cs
@@ -7,6 +7,20 @@ namespace System
 {
     public static partial class Environment
     {
-        public static unsafe long WorkingSet => Interop.Process.GetProcInfo(ProcessId, true, out _)->ki_rssize;
+        public static unsafe long WorkingSet
+        {
+            get
+            {
+                Interop.Process.kinfo_proc* processInfo = Interop.Process.GetProcInfo(ProcessId, true, out _);
+                try
+                {
+                    return processInfo->ki_rssize;
+                }
+                finally
+                {
+                    NativeMemory.Free(processInfo);
+                }
+            }
+        }
     }
 }

